src/Agent.Listener/Configuration/AutoLogonRegistryManager.cs

```diff
@@ -0,0 +1,339 @@
+#if OS_WINDOWS
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.Win32;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Security.Principal;
+
+namespace Microsoft.VisualStudio.Services.Agent.Listener.Configuration
+{
+    [ServiceLocator(Default = typeof(AutoLogonRegistryManager))]
+    public interface IAutoLogonRegistryManager : IAgentService
+    {
+        void LogWarnings(string domainName, string userName);
+        void UpdateRegistrySettings(string domainName, string userName);
+        bool DoesRegistryExistForUser(string domainName, string userName);
+        void RevertRegistrySettings(string domainName, string userName);
+    }
+
+    public class AutoLogonRegistryManager : AgentService, IAutoLogonRegistryManager
+    {
+        private IWindowsRegistryManager _registryManager;
+        private INativeWindowsServiceHelper _windowsServiceHelper;
+
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+            _registryManager = hostContext.GetService<IWindowsRegistryManager>();
+            _windowsServiceHelper = hostContext.GetService<INativeWindowsServiceHelper>();
+        }
+
+        public void LogWarnings(string domainName, string userName)
+        {
+            var warningReasons = GetWarningsForMachineSpecificSettings();
+            warningReasons.AddRange(GetWarningsForUserSpecificSettings(domainName, userName));
+
+            if (warningReasons.Count > 0)
+            {
+                var terminal = HostContext.GetService<ITerminal>();
+                terminal.WriteLine();
+                terminal.WriteLine(StringUtil.Loc("UITestingWarning"));
+                for (int i=0; i < warningReasons.Count; i++)
+                {
+                    terminal.WriteLine(String.Format("{0} - {1}", i + 1, warningReasons[i]));
+                }
+                terminal.WriteLine();
+            }
+        }
+
+        public void UpdateRegistrySettings(string domainName, string userName)
+        {
+            ShowAutoLogonWarningIfAlreadyEnabled(domainName, userName);
+
+            //machine specific
+            UpdateMachineSpecificRegistrySettings(domainName, userName);
+
+            //user specific
+            UpdateUserSpecificRegistrySettings(domainName, userName);
+        }
+
+        public void RevertRegistrySettings(string domainName, string userName)
```

Comment:
if the latest value is not what we had set, I am not touching.

---
In reply to: [118602034](https://github.com/Microsoft/vsts-agent/pull/985#discussion_r118602034) [](ancestors = 118602034)